config/v1/types_infrastructure.go

```diff
@@ -706,6 +701,55 @@ type GCPResourceTag struct {
 	Value string `json:"value"`
 }
 
+// CloudLoadBalancerConfig contains IPs for API, API-Int and Ingress Cloud Load Balancers
+// +union
+// +kubebuilder:validation:XValidation:rule="has(self.dnsType) && self.dnsType == 'ClusterHosted' ? has(self.clusterHosted) && size(self.clusterHosted.apiIntLoadBalancerIPs) > 0 : true",message="when dnsType is ClusterHosted, API-Int Load Balancer IP addresses have to specified with clusterHostedDNS"
+type CloudLoadBalancerConfig struct {
+	// dnsType indicates the type of DNS solution in use within the cluster. Its default value of
+	// `PlatformDefault` indicates that the cluster's DNS is the default provided by the cloud platform.
+	// It can be set to `ClusterHosted` to bypass the configuration of the cloud default DNS. In this mode,
+	// the cluster needs to provide a self-hosted DNS solution for the cluster's installation to succeed.
+	// The cluster's use of the cloud's Load Balancers is unaffected by this setting.
+	// The value is immutable after it has been set at install time.
+	// Currently, there is no way for the customer to add additional DNS entries into the cluster hosted DNS.
+	// Enabling this functionality allows the user to start their own DNS solution outside the cluster after
+	// installation is complete. The customer would be responsible for configuring this custom DNS solution,
+	// and it can be run in addition to the in-cluster DNS solution.
+	// +default="PlatformDefault"
+	// +kubebuilder:validation:Enum="ClusterHosted";"PlatformDefault"
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="dnsType is immutable"
+	// +optional
+	// +unionDiscriminator
+	dnsType DNSType `json:"dnsType,omitempty"`
+
+	// clusterHosted holds the IP addresses of API, API-Int and Ingress Load
+	// Balancers on Cloud Platforms. The DNS solution hosted within the cluster
+	// use these IP addresses to provide resolution for API, API-Int and Ingress
+	// services.
+	// +optional
+	ClusterHosted *CloudLoadBalancerIPs `json:"clusterHosted,omitempty"`
+}
+
+type CloudLoadBalancerIPs struct {
+	// apiIntLoadBalancerIPs holds Load Balancer IPs for API-Int.
+	// +kubebuilder:validation:Format=ip
+	// +listType=set
+	ApiIntLoadBalancerIPs []IP `json:"apiIntLoadBalancerIPs"`
```

Comment:
Is there a length limit we can set on these lists?

Reply:
No. The user has the ability to bring in their own subnets. That may mean an additional LB on that subnet and so additional IPs. And then if we also consider dual-stack clusters, then it is hard to predict how many IP addresses could make it into this list.

Reply:
I think we can still limit this to some sensible, but still permissive enough limit, 16, 32, maybe 64? But I don't think we can leave this unbounded. No one is going to add hundreds of API LB IPs right?